### PR TITLE
Json formatter

### PR DESCRIPTION
**Pull request by Crowy-o**
Created class JSONFormatter. How to use: creating an inhereted class (such as JSONFormaterLogicalDisk). This class will be sent to JSONFormatSaver. It could save in 2 ways: value of date is an array of values (TryAddJSONFormattedData()) and value of date is 1 value (TrySetJSONFormattedData()). 
Also created namespace Utilities in order to create fucntion GetCurrentDateAndTimeFormatted(). It returns current time in system in the format: dd.mm.yyyy hh:mm:ss (example: 12.01.2021 19:58:32). Added inside Utilities namespace HelperDateAndTime in order to separate fucntion, that must work only with GetCurrentDateAndTimeFormatted from other functions, that might be added in further

### CODE REVIEW

**Review by AntonVoytov — CHANGES_REQUESTED** [review]
Rework according to meeting

File: Server/BackEndMonitoringServer/include/stdafx.h [hunk 1]
```diff
@@ -1,3 +1,9 @@
 #pragma once
 
-#include <iostream>
\ No newline at end of file
+#include <iostream>
+#include <fstream>
+#include <chrono>
+#include <ctime>
+#include <iomanip>
+#include "json.hpp"
```
Comment (AntonVoytov) [inline Server/BackEndMonitoringServer/include/stdafx.h:8]: Move "json.hpp" to files where JSON is used.

File: Server/BackEndMonitoringServer/src/BackEndMonitoringServer.cpp [hunk 2]
```diff
@@ -1,26 +1,6 @@
-﻿// BackendMonitoring.cpp : Этот файл содержит функцию "main". Здесь начинается и
-// заканчивается выполнение программы.
-//
+﻿#include "stdafx.h"
```
Comment (AntonVoytov) [inline Server/BackEndMonitoringServer/src/BackEndMonitoringServer.cpp:1]: We include stdafx 2 times

File: utils/include/CJSONFormatSaver.h [hunk 3]
```diff
@@ -0,0 +1,22 @@
+#pragma once
+
+#include "stdafx.h"
```
Comment (AntonVoytov) [inline utils/include/CJSONFormatSaver.h:3]: Remove includes from here they are already present in cpp files

File: utils/include/CJSONFormatSaver.h [hunk 4]
```diff
@@ -0,0 +1,22 @@
...
+#include "CJSONFormatter.h"
+
+class CJSONFormatSaver
+{
+public:
+	CJSONFormatSaver() = delete;
+	CJSONFormatSaver(std::string path_to_file) :
```
Comment (AntonVoytov) [inline utils/include/CJSONFormatSaver.h:10]: const string &

File: utils/include/CJSONFormatSaver.h [hunk 5]
```diff
@@ -0,0 +1,22 @@
...
+		m_path_to_file_(path_to_file)
+	{ };
+	CJSONFormatSaver(CJSONFormatSaver&) = delete;
+	CJSONFormatSaver(CJSONFormatSaver&&) = delete;
+	
+	bool TrySaveToFile(CJSONFormatter& formatted_data);
+
+private:
+	std::string m_path_to_file_;
+	const unsigned short m_number_of_spaces_ = 3;
```
Comment (AntonVoytov) [inline utils/include/CJSONFormatSaver.h:20]: initialize this const variable in the constructor

File: utils/include/CJSONFormatSaver.cpp [hunk 6]
```diff
@@ -0,0 +1,36 @@
...
+		json_file_to_read >> buff;
+	}
+	json_file_to_read.close();
+
+	if (buff.is_null())
+	{
+		buff.merge_patch(*formatted_data.GetJSONFormattedData());
+	}
+
+
```
Comment (AntonVoytov) [inline utils/include/CJSONFormatSaver.cpp:24]: too many empty lines

File: utils/include/CJSONFormatSaver.cpp [hunk 7]
```diff
@@ -0,0 +1,36 @@
...
+    std::ofstream JSON_file_to_save(m_path_to_file_);
```
Comment (AntonVoytov) [inline utils/include/CJSONFormatSaver.cpp:25]: add whitespace

File: utils/include/CJSONFormatSaver.cpp [hunk 8]
```diff
@@ -0,0 +1,36 @@
...
+
+	if (!JSON_file_to_save.is_open())
+	{
+		return false;
+	}
+
+
```
Comment (AntonVoytov) [inline utils/include/CJSONFormatSaver.cpp:32]: too many empty spaces

File: utils/include/CJSONFormatter.cpp [hunk 9]
```diff
@@ -0,0 +1,38 @@
+#include "stdafx.h"
+#include "CJSONFormatter.h"
+
+bool CJSONFormatter::TryAddJSONFormattedData(const nlohmann::json* 
+    p_formatted_data)
+{
+    if(p_formatted_data->is_null())
```
Comment (AntonVoytov) [inline utils/include/CJSONFormatter.cpp:7]: if( -> if (

File: utils/include/CJSONFormatter.h [hunk 10]
```diff
@@ -0,0 +1,22 @@
+#pragma once
+
+#include "stdafx.h"
```
Comment (AntonVoytov) [inline utils/include/CJSONFormatter.h:3]: Remove includes

File: utils/include/ThreadSafeVariable.h [hunk 11]
```diff
@@ -0,0 +1,32 @@
+#pragma once
+
+#include <mutex>
+
+template <typename T>
+class ThreadSafeVariable
```
Comment (AntonVoytov) [inline utils/include/ThreadSafeVariable.h:6]: Fix this file according to code style, or it was already fixed just pull those changes

File: utils/include/Utilities.cpp [hunk 12]
```diff
@@ -0,0 +1,61 @@
+#include "stdafx.h"
+#include "Utilities.h"
+
+std::string Utilities::GetCurrentDateAndTimeFormatted()
```
Comment (AntonVoytov) [inline utils/include/Utilities.cpp:4]: bool Utilities::GetCurrentDateAndTimeFormatted(string &)

File: utils/include/Utilities.cpp [hunk 13]
```diff
@@ -0,0 +1,61 @@
...
+{
+    //parse to format: dd.mm.yyyy hh:mm:ss
+    std::time_t current_time = std::chrono::system_clock::to_time_t(
+        std::chrono::system_clock::now());
+    std::string date_and_time = ctime(&current_time);
+
+    date_and_time.pop_back();
+    date_and_time = date_and_time.substr(date_and_time.find_first_of(" ") + 1);
```
Comment (AntonVoytov) [inline utils/include/Utilities.cpp:12]: +1 -> +1ULL

File: utils/include/Utilities.cpp [hunk 14]
```diff
@@ -0,0 +1,61 @@
...
+    }
+    
+    date_and_time.insert(date_and_time.find_first_of(" "), "." + buff_month);
+    date_and_time.insert(date_and_time.find_first_of(" "), 
+        "." + buff_year + " ");
+
+    return date_and_time;
+}
+
+bool Utilities::HelperDateAndTime::TrySetMonthAsNumber(std::string* p_month)
```
Comment (AntonVoytov) [inline utils/include/Utilities.cpp:36]: if you are passing a pointer to a function you should check it if it not nullptr

File: utils/include/Utilities.h [hunk 15]
```diff
@@ -0,0 +1,18 @@
+#pragma once
+
+#include "stdafx.h"
```
Comment (AntonVoytov) [inline utils/include/Utilities.h:3]: move to cpp

File: utils/include/Utilities.h [hunk 16]
```diff
@@ -0,0 +1,18 @@
...
+
+namespace Utilities
+{
+	/// <summary>
+	/// Return current date and time in system
+	/// </summary>
+	/// <returns>format: dd.mm.yyyy hh:mm:ss</returns>
+	std::string GetCurrentDateAndTimeFormatted();
+
+	namespace HelperDateAndTime
```
Comment (AntonVoytov) [inline utils/include/Utilities.h:13]: No needs for nested namespaces, delete this namespace or separate it from Utilities

File: utils/include/Utilities.h [hunk 17]
```diff
@@ -0,0 +1,18 @@
+#pragma once
+
+#include "stdafx.h"
+
+namespace Utilities
```
Comment (AntonVoytov) [inline utils/include/Utilities.h:5]: Maybe we should rename the namespace to Utils?

**Review by AntonVoytov — COMMENTED** [review]
Please fix comments and merge the code to the main.
1). Do we need those binary files? If not remove them.
2). In the future JSON formatter will be reworked

**Review by AntonVoytov — APPROVED** [review]

File: utils/include/CJSONFormatSaver.cpp (hunk 7 above)
Comment (Crowy-o) [reply utils/include/CJSONFormatSaver.cpp:25]: I don't get what you mean. Like this:  std::ofstream JSON_file_to_save( m_path_to_file_ ); ?

**Pull request merged**